<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME =
USER =
PASSWORD =
HOST =
PORT =

SyntaxError: invalid syntax (1531360337.py, line 2)

In [ ]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [ ]:
# текст запроса
query_3_1 = f'''
select
    count(v.id) as count_vac
from
    public.vacancies as v
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count_vac
0      49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [ ]:
# текст запроса
query_3_2 = f'''
select
    count(e.id) count_emp
from
    public.employers as e
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count_emp
0      23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
# текст запроса
query_3_3 = f'''
select
    count(a.id) as count_ar
from
    public.areas as a
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count_ar
0      1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
# текст запроса
query_3_4 = f'''
select
    count(i.id) count_ind
from
    public.industries as i
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count_ind
0        294

***

Выводы по предварительному анализу данных:
1. Количество вакансий в базе данных составляет - 49197. Вакансий в базе в 2 раза больше, чем работадателей, т.е. в среднем 2 вакансии на 1 работадателя.
2. Количество работодателей в базе данных составляет - 23501. Работадателей в базе в 2 раза меньше, чем вакансий, т.е. в среднем 1 работадатель на 2 вакансии.
3. Количество регионов (городов) в базе данных составляет - 1362. Представлены практически все основные города России и ближнего зарубежья.
4. Количество сфер деятельности в базе данных составляет - 294.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
n = 5
query_4_1 = f'''
select
    count(v.id) AS cnt,
    a.name AS area 
from
    public.vacancies AS v
    join public.areas AS a on v.area_id = a.id
group by 2
order by 1 desc
limit {n}
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''
select
    count(v.id) as cnt_vac
from
    public.vacancies as v
where
    salary_from is not null
    or salary_to is not null
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt_vac
0    24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
# текст запроса
query_4_3 = f'''
select
    round(AVG(v.salary_from), 0) as sai_f,
    round(AVG(v.salary_to), 0) as sal_t
from
    public.vacancies as v
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


sai_f     sal_t
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''
select
    count(v.id) as cnt_vac_s_e,
    v.schedule,
    v.employment
from
    public.vacancies as v
group by 2, 3
order by 1 desc
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


cnt_vac_s_e          schedule           employment
0         35367       Полный день     Полная занятость
1          7802  Удаленная работа     Полная занятость
2          1593     Гибкий график     Полная занятость
3          1312  Удаленная работа  Частичная занятость
4           940    Сменный график     Полная занятость
5           569       Полный день           Стажировка
6           367    Вахтовый метод     Полная занятость
7           347       Полный день  Частичная занятость
8           312     Гибкий график  Частичная занятость
9           141       Полный день     Проектная работа
10          133  Удаленная работа     Проектная работа
11          116     Гибкий график           Стажировка
12          101    Сменный график  Частичная занятость
13           64  Удаленная работа           Стажировка
14           18     Гибкий график     Проектная работа
15           12    Сменный график           Стажировка
16            2    Вахтовый метод     Проектная работа
17            1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
# текст запроса
query_4_5 = f'''
select
    experience as "Требуемый опыт работы",
    count(v.id) as cnt_vac_exp
from
    public.vacancies as v
group by 1
order by 2
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы  cnt_vac_exp
0           Более 6 лет         1337
1             Нет опыта         7197
2         От 3 до 6 лет        14511
3    От 1 года до 3 лет        26152

***

Выводы по детальному анализу вакансий:
1. Наибольшее количество вакансий находится в крупных городах-миллионниках. Безусловным лидером является Москва - 5333 вакансий.
2. Почти у половины вакансии определён уровень зарплаты - 24073 из 49197. Зарплаты остальных вакансий видимо определяются по результатам собеседования.
3. Средние значения для нижней и верхней границы зарплатной вилки составляют: 71065 и 110537 руб. соответственно. Эти показатели выше средних зарплат в России и ближнем зарубежье.
4. Наиболее востребовательными вариантами сочетания типа рабочего графика и типа трудоустройства, используемого в вакансиях являются: "Полный день и Полная занятость" - 35367 вакансий; на втором месте (с большим отрывом) "Удаленная работа и Полная занятость" - 7802 вакансий.
5. Наиболее востребованный опыт работы "От 1 года до 3 лет" - 26152 вакансий. Это может быть связано с одной стороны с ожиданием требований у претендентов с низким опытом работы более низкой заплаты, чем претендентов с опытом работы "Более 6 лет". С другой стороны опыт работы "От 1 года до 3 лет" не требует больших вложений в переобучение, в отличии от претендентов у которых "Нет опыта" работы.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query_5_1 = f'''
select
    e.name as name_employers,
    count(v.id) as cnt_vac
from
    public.employers as e
    left join public.vacancies as v on v.employer_id = e.id
group by 1
order by 2 desc
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name_employers  cnt_vac
0                                  Яндекс     1933
1                              Ростелеком      491
2                                Тинькофф      444
3                                    СБЕР      428
4                           Газпром нефть      331
...                                   ...      ...
23170      Капитал Лайф Страхование Жизни        0
23171                            AGASTONE        0
23172                       Независимость        0
23173  Верховный Суд Республики Татарстан        0
23174                            Трекмарк        0

[23175 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query_5_2 = f'''
select 
    a.name as name_area,
    count(distinct e.id) as cnt_emp,
    count(distinct v.id) as cnt_vac
from
    public.areas a
    left join public.employers e on e.area = a.id
    left join public.vacancies v on v.area_id = a.id
where v.id is null
group by a.name
order by 2 desc
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name_area  cnt_emp  cnt_vac
0                            Россия      410        0
1                         Казахстан      207        0
2                Московская область       75        0
3                Краснодарский край       19        0
4                          Беларусь       18        0
..                              ...      ...      ...
588                        Кувандык        0        0
589                        Куйбышев        0        0
590  Куйбышево (Ростовская область)        0        0
591                        Култаево        0        0
592                        Кульсары        0        0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''
select 
    e.name as name_e,
    count(distinct v.area_id) as cnt_ar_v
from
    public.employers as e
    join public.vacancies as v on v.employer_id = e.id
group by 1
order by 2 desc
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name_e  cnt_ar_v
0                      Яндекс       181
1                  Ростелеком       152
2                  Спецремонт       116
3      Поляков Денис Иванович        88
4                   ООО ЕФИН         71
...                       ...       ...
14761                  UniSol         1
14762            UNISTORY LLC         1
14763                   UNIT6         1
14764     United Distribution         1
14765              UNITEL LLC         1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# текст запроса
query_5_4 = f'''
select
    count(e.name) as name_e
from
    public.employers as e
    left join public.employers_industries as ei on ei.employer_id = e.id
where industry_id is null

            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


name_e
0    8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# текст запроса
query_5_5 = f'''
select
    e.name as name_e,
    count(i.name) as name_i 
from
    public.employers_industries as ei
    join public.employers as e on e.id = ei.employer_id
    join public.industries as i on i.id = ei.industry_id
group by 1
having count(i.name) = 4
order by 1
limit 3
           '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name_e  name_i
0  101 Интернет       4
1      21vek.by       4
2          2ГИС       4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# текст запроса
query_5_6 = f'''
select
    count(i.name) as name_i 
from
    public.employers_industries as ei
    join public.employers as e on e.id = ei.employer_id
    join public.industries as i on i.id = ei.industry_id
where i.name = 'Разработка программного обеспечения'
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


name_i
0    3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников

# Импорт библиотек
import pandas as pd # библиотека для анализа данных
import requests # библиотека для обработки запросов
from bs4 import BeautifulSoup # библиотека для анализа HTML-документов

# Запрос HTML-ответа с использованием URL-адреса и получение ответа в виде html
url="https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров"
table_class="standard sortable jquery-tablesorter"
response=requests.get(url)

# Преобразование данных из html в объект beautifulsoup
page = BeautifulSoup(response.text, 'html.parser')
rustable=page.find('table',{'class':"standard"})

# Преобразование таблицы Википедии в фрейм данных Python
df=pd.read_html(str(rustable))

# Преобразование списка в фрейм данных
df=pd.DataFrame(df[0])

# Удаление ненужных столбцов
data = df.drop(["№", "Прирост  относительно  переписи-2010", "Население, тыс. чел. (итоги переписи  2010 года)[6]"], axis=1)

# Переименование столбцов для удобства
data = data.rename(columns={"Население, тыс. чел.  (итоги переписи  2021 года)[5]": "Население"})

# Изменим значение для ячейки Воронеж[a] на Воронеж
data.replace("Воронеж[a]", "Воронеж", inplace=True)

# Выгружаем данные из DataFrame в CSV-файл и сохраняем файл в папке
data.to_csv('cities.csv', index=False, sep=' ')

# Преобразование столбца "Город" в список
city_list = list(data['Город'])
print(city_list)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград']


In [ ]:
# текст запроса
query_5_7 = f'''
select
	a.name,
   	count(*) cnt
from
	public.employers e
	join public.vacancies v on e.id = v.employer_id
	join public.areas a on a.id = v.area_id
where
	e.name = 'Яндекс'
  	and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
                 	'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск',
                 	'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар',
                       'Воронеж',  'Пермь', 'Волгоград')
group by a.id

union all

select
	'Total',
    count(*)
from
	public.vacancies v
	join public.employers e on e.id = v.employer_id
	join public.areas a on a.id = v.area_id
where
	e.name = 'Яндекс'
  	and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
                 	'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск',
                 	'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар',
                       'Воронеж',  'Пермь', 'Волгоград')
order by cnt
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7            Самара   26
8               Уфа   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

Выводы по анализу работодателей:
1. Наибольшее число вакансий среди работодателей принадлежит "Яндексу" - 1933, это большой отрыв от находящегося на втором месте "Ростелекома" - 491.
2. Среди регионов наибольшее число вакансий у России (410) и Казахстана (207), отсутствуют вакансии в небольших городах.
3. "Яндекс" и "Ростелеком", как и ожидалось, обладают наибольшим числом регионов с вакансиями, 181 и 152 соответственно.
4. Довольно много работодателей не указало сферу своей деятельности 8419 из 23175.
5. У многих компаний указано несколько сфер деятельности (четыре сферы деятельности более чем у 1000 компаний).
6. Более 3,5 тысяч компаний указали сферу деятельности "Разработка программного обеспечения".
7. Наибольшее количество вакансий в крупных городах. Так по "Яндексу" в Москве - 54 вакансии, в Санкт-Петербурге - 42 вакансии.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса
query_6_1 = f'''
select
    count(v.name) as name_v
from
    public.vacancies as v
where
    v.name ilike '%data%'
    or v.name ilike '%данн%'
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_9536\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


name_v
0    1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса
query_6_2 = f'''
select
	count(*) cnt
from
	public.vacancies as v
where
	(name ilike '%data scientist%'
   	or name ilike '%data science%'
   	or name ilike '%исследователь данных%'
   	or (name like '%ML%' and name not ilike '%HTML%')
   	or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%')
 	and (name ilike '%junior%'
  	or v.experience = 'Нет опыта'
  	or v.employment='Стажировка')
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_9536\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = f'''
select
	count(*) cnt
from
	public.vacancies as v
where
	(name ilike '%data scientist%'
   	or name ilike '%data science%'
   	or name ilike '%исследователь данных%'
   	or (name like '%ML%' and name not ilike '%HTML%')
   	or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%')
   	and (v.key_skills ilike '%postgres%'
    or v.key_skills ilike '%SQL%')
 	'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_9536\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


cnt
0  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = f'''
select
	count(*) cnt
from
	public.vacancies as v
where
	(name ilike '%data scientist%'
   	or name ilike '%data science%'
   	or name ilike '%исследователь данных%'
   	or (name like '%ML%' and name not ilike '%HTML%')
   	or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%')
   	and v.key_skills ilike '%Python%'
    '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_9536\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


cnt
0  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
query_6_5 = f'''
select
    round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2) as avg_ks
from
	public.vacancies
where
	(name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name like '%ML%'
    and name not like '%HTML%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%')
		    '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_9536\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg_ks
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса

query_6_6 = f'''
select distinct
    experience,
    round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to)), 0) as mean_salary
from
    public.vacancies
where
    (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name like '%ML%'
    and name not like '%HTML%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%')
    and (salary_from is not null
    or salary_to is not null)
group by 1
           '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_9536\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  mean_salary
0           Нет опыта      74643.0
1  От 1 года до 3 лет     139675.0
2       От 3 до 6 лет     243115.0

***

Выводы по предметному анализу:
1. Имеет отношение к данным 1771 вакансия.
2. Для начинающего дата-сайентиста имеется 51 подходящия вакансия.
3. Имеется 201 вакансия для DS, в которых в качестве ключевого навыка указан SQL или postgres.
4. Python довольно популярен в требованиях работодателей к DS. Количество вакансий, в которых в качестве ключевого навыка указан Python - 351.
5. В среднем указывают в вакансиях для DS 6,41 ключевых навыков.
6. Для специалистов DS в среднем указывают довольно высокую зарплату. Чем выше опыт, тем более высокая зарплата. Для опыта работы от 3 до 6 лет в среднем около 243 тысяч руб.


# Общий вывод по проекту

Подведем итог исследования, обобщим выводы.
Здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования.
1. Всего вакансий - 49197. Вакансий, имеющих отношение к данным - 1771. Больше всего вакансий: Системный администратор (1761), Программист 1С (1153), Бизнес-аналитик (488), Специалист технической поддержки (457) и Системный аналитик (432).
2. Всего в таблице редставлено 1362 регионов. Наибольшее количество работодателей представлено в следующих регионах: Россия (410), Казахстан (207) и Московская область (75). Наибольшее количество вакансий представлено в следующих городах: Москва (5333), Санкт-Петербург (2851) и Минск (2112). Эти данные говорят о том что, чем больше регион (город), тем больше в нём представлено вакансий и работодателей.
3. На основании анализа данных можно сделать заключение, что специалисты в IT-отрасли очень востребованы на рынке труда.
4. Было бы интересно провести анализ не только вакансий, но и резюме соискателей. Сравнить предложения вакансий работодателей и запросов соискателей.

In [ ]:
# Определим каких вакансий больше всего
query_7_1 = f'''
select
    v.name,
    count(v.id) as count_vac
from
    public.vacancies as v
group by 1
order by 2 desc
            '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Александр\AppData\Local\Temp\ipykernel_6364\2931715827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


name  count_vac
0                     Системный администратор       1761
1                              Программист 1С       1153
2                             Бизнес-аналитик        488
3            Специалист технической поддержки        457
4                          Системный аналитик        432
...                                       ...        ...
21218            Генеральный менеджер Проекта          1
21219    Ведущий инженер-программист (1С:УПП)          1
21220       Главный специалист IT направления          1
21221  Сотрудник Отдела информационных систем          1
21222             Старший финансовый менеджер          1

[21223 rows x 2 columns]

# ФИНАЛ

In [1]:
# не забываем закрыть соединение после окончания работы
connection.close()

NameError: name 'connection' is not defined